File: Binary Search/Hard/LET_no1891_Cutting Ribbons.py
class Solution:
    def maxLength(self, ribbons: List[int], k: int) -> int:
        left, right = 1, max(ribbons)
        while left + 1 < right:
            mid = (left + right) // 2
            if self.helper(ribbons, mid) >= k:
                left = mid
            else:
                right = mid
        
        if self.helper(ribbons, right) >= k:
            return right
        if self.helper(ribbons, left) >= k:
            return left
        return 0
    
    def helper(self, arr, target):
        return sum(num // target for num in arr)

File: Binary Search/Hard/LET_no1891_Cutting Ribbons.cpp
class Solution {
public:
    int maxLength(vector<int>& ribbons, int k) {
        int min_val = 1, max_val = *max_element(ribbons.begin(), ribbons.end());
        while (min_val + 1 < max_val) {
            int mid_val = min_val + (max_val - min_val) / 2;
            
            if (getPieces(ribbons, mid_val) >= k)
                min_val = mid_val;
            else
                max_val = mid_val;
        }

        if (getPieces(ribbons, max_val) >= k)
            return max_val;
        if (getPieces(ribbons, min_val) >= k)
            return min_val;
        return 0;
    }

    int getPieces(vector<int>& ribbons, int length) {
        int pieces = 0;
        for (int l : ribbons)
            pieces += l / length;
        return pieces;
    }
};